File: solutions/1816-lowest-common-ancestor-of-a-binary-tree-iv/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode[] nodes) {
        ArrayList<Integer> nodeValues = new ArrayList<>();
        for (TreeNode node : nodes) {
            nodeValues.add(node.val);
        }
        
        ArrayList<TreeNode> ancestors = new ArrayList<>();
        findMatchingNodes(root, nodeValues, ancestors);
        
        return ancestors.isEmpty() ? null : ancestors.get(0);
    }
    
    private int findMatchingNodes(TreeNode root, ArrayList<Integer> targetValues, ArrayList<TreeNode> ancestors) {
        if (root == null) {
            return 0;
        }
        
        int matchCount = 0;
        
        // Count matching nodes in left and right subtrees
        matchCount += findMatchingNodes(root.left, targetValues, ancestors);
        matchCount += findMatchingNodes(root.right, targetValues, ancestors);
        
        // Check if current node is a target
        if (targetValues.contains(root.val)) {
            matchCount++;
        }
        
        // If we've found all target nodes, this could be the LCA
        if (matchCount == targetValues.size()) {
            ancestors.add(root);
        }
        
        return matchCount;
    }
}

File: solutions/1816-lowest-common-ancestor-of-a-binary-tree-iv/solution.py
# Definition for a binary tree node
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', nodes: 'List[TreeNode]') -> 'TreeNode':
        node_values = [node.val for node in nodes]
        ancestors = []
        
        def find_matching_nodes(root: TreeNode) -> int:
            if not root:
                return 0
                
            match_count = 0
            match_count += find_matching_nodes(root.left)
            match_count += find_matching_nodes(root.right)

            if root.val in node_values:
                match_count += 1

            if match_count == len(node_values):
                ancestors.append(root)
                
            return match_count
            
        find_matching_nodes(root)

        return ancestors[0] if ancestors else None


